src/runtime-rs/crates/kata/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid argument")]
+    InvalidArgument,
+    #[error("failed to get system time: {0}")]
+    SystemTime(#[source] std::time::SystemTimeError),
+    #[error("error from sandbox: {0}")]
+    Sandbox(#[source] virtcontainers::Error),
+    #[error("failed to get bundle path: {0}")]
+    BundlePath(#[source] std::io::Error),
+    #[error("failed to load oci spec: {0}")]
+    OciSpec(#[source] oci_spec::OciSpecError),
+    #[error("failed to open file {1} with error: {0}")]
+    OpenFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("failed to get metadata of file {1} with error: {0}")]
+    QueryFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("failed to write file {1} with error: {0}")]
+    WriteFile(#[source] std::io::Error, std::path::PathBuf),
+    #[error("empty sandbox id")]
+    EmptySandboxId,
+    #[error("failed to extract oci spec information: {0}")]
+    OciSpecInfo(#[source] virtcontainers::spec_info::OciSpecInfoError),
+    #[error("failed to spawn child: {0}")]
+    SpawnChild(#[source] std::io::Error),
+    #[error("failed to bind socket at {1} with error: {0}")]
+    BindSocket(#[source] std::io::Error, PathBuf),
+    #[error("failed to get self exec: {0}")]
+    SelfExec(#[source] std::io::Error),
+    #[error("error from config: {0}")]
+    Config(#[source] config::ConfigError),
+    #[error("failed to set logger")]
+    Logger,
+    #[error("failed to start ttrpc server: {0}")]
+    StartServer(#[source] ttrpc::error::Error),
+    #[error("failed to wait ttrpc server when {0}")]
+    WaitServer(String),
+    #[error("failed to get env variable: {0}")]
+    EnvVar(#[source] std::env::VarError),
+    #[error("failed to parse server fd {0}")]
+    ServerFd(String),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+const SOCKET_ROOT: &str = "/run/containerd";
+const SHIM_SOCKET: &str = "shim.sock";
+const KATA_BIND_FD: &str = "KATA_BIND_FD";
+
+/// Received commandline arguments or environment arguments.
+///
+/// For defail information, please refer to the
+/// [shim v2 spec](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md).
+#[derive(Debug, Default, Clone)]
+pub struct ShimArgs {
+    /// the id of the container
+    pub id: String,
+    /// the namespace for the container
+    pub namespace: String,
+    /// the address of the containerd's main socket
+    pub address: String,
+    /// the binary path to publish events back to containerd
+    pub publish_binary: String,
+    /// the path to the bundle to delete
+    pub bundle: String,
+    /// Whether or not to enable debug
+    pub debug: bool,
+}
+
+impl ShimArgs {
+    /// Check the shim argument object is vaild or not.
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, is_delete: bool) -> Result<()> {
```

Comment:
@jodh-intel I _think_ `is_delete` is meant to signal to `validate()` whether or not the validation is to be performed for a delete Action, because if so, `validate()` needs also to validate the bundle arg which isn't required otherwise.

But let's wait for an authoritative reply. :-)